File: OnePwave/FullSpace/getPWaveAmp.m
function [ampOut] = getPWaveAmp(x,y,z,vP,f,thetaW,phiW)
% for a given y,z and x grids
% get the Plane P-wave amplitude
% x, y, z are column vectors, z are scalar values
% vP is the P-wave velocity in m/s, f is frequency in Hz
% thetaW and phiW are the dierction of the plane wave
lambda = vP/f;
k = 2*pi/lambda;
kVec = [k*sind(thetaW)*cosd(phiW);k*sind(thetaW)*sind(phiW);k*cosd(thetaW)];
rVec = [x,y,z];
i = sqrt(-1);
kr = rVec*kVec;
ampOut = exp(-i*(kr));
kUnitVec = kVec./(sqrt(kVec(1,1)^2+kVec(2,1)^2+kVec(3,1)^2));
ampOut = [ampOut*kUnitVec(1,1),ampOut*kUnitVec(2,1),ampOut*kUnitVec(3,1)];

end

File: OnePwave/HalfSpace/getPWaveAmp.py
def getPWaveAmp(x, y, z, vP, f, thetaW, phiW):
    
    lambda_ = vP / f
    k = 2 * np.pi / lambda_

    kVec = np.array([
        k * np.sin(np.deg2rad(thetaW)) * np.cos(np.deg2rad(phiW)),
        k * np.sin(np.deg2rad(thetaW)) * np.sin(np.deg2rad(phiW)),
        k * np.cos(np.deg2rad(thetaW))
    ])

    rVec = np.column_stack((x, y, z))

    kr = rVec @ kVec
    amp = np.exp(-1j * kr)
#normalize
    kUnitVec = kVec / np.sqrt(np.sum(kVec**2))

    
    ampOut = np.column_stack((
        amp * kUnitVec[0],
        amp * kUnitVec[1],
        amp * kUnitVec[2]
    ))

    return ampOut
